components/heap/heap_caps.c

```diff
@@ -77,6 +85,54 @@ esp_err_t heap_caps_register_failed_alloc_callback(esp_alloc_failed_hook_t callb
     return ESP_OK;
 }
 
+IRAM_ATTR static bool heap_caps_low_memory_check( size_t size, uint32_t caps)
+{
+#if CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        heap_low_memory_callback_t ith = low_memory_callbacks[i];
+        if (ith.callback && (ith.caps & caps) > 0) {
+            size_t free = heap_caps_get_free_size(ith.caps);
+            if (free < ith.threshold + size) {
+                return ith.callback(caps, size, free);
+            }
+        }
+    }
+#endif
+    return true;
+}
+
+esp_err_t heap_caps_register_low_heap_callback(uint32_t caps, uint32_t threshold, esp_heap_low_memory_hook_t callback)
+{
+    if (callback == NULL) {
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    // exists?
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        if (low_memory_callbacks[i].caps == caps) {
+            low_memory_callbacks[i].caps = caps;
+            low_memory_callbacks[i].threshold = threshold;
+            low_memory_callbacks[i].callback = callback;
+            return ESP_OK;
+        }
+    }
+
+    // add
+    for (int i = 0; i < CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT; i++) {
+        if (low_memory_callbacks[i].callback == NULL) {
+            low_memory_callbacks[i].caps = caps;
+            low_memory_callbacks[i].threshold = threshold;
+            low_memory_callbacks[i].callback = callback;
+            return ESP_OK;
+        }
+    }
+
+    return ESP_FAIL;
+}
+
+#if CONFIG_HEAP_LOW_MEMORY_CALLBACK_COUNT
+IRAM_ATTR
+#endif
```

Comment:
I don't see the reason why adding placing this function in IRAM if callback count > 0.

Reply:
`heap_caps_match` is used by `heap_caps_get_free_size` which is now called in all of the allocators to check if the threshold is breached.

Reply:
It makes sense, I didn't see that you also added the `IRAM_ATTR` to `heap_caps_get_free_size()`